content/blog/2023-10-18-skywalking-toolkit-trace/index.md

```diff
@@ -0,0 +1,285 @@
+	defer trace.StopSpan()
+
+	// Prepare to send an HTTP request.
+	resp, err := getProvider()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return
+	}
+	_, _ = w.Write(body)
+}
+
+func main() {
+	http.HandleFunc("/consumer", consumerHandler)
+
+	_ = http.ListenAndServe(":9999", nil)
+}
+```
+
+```go
+// provider.go
+package main
+
+import (
+	"net/http"
+
+	_ "github.com/apache/skywalking-go"
+	"github.com/apache/skywalking-go/toolkit/trace"
+)
+
+func providerHandler(w http.ResponseWriter, r *http.Request) {
+	//Create an EntrySpan to trace the execution of the providerHandler method.
+	trace.CreateEntrySpan("GET:/provider", func(headerKey string) (string, error) {
+		return r.Header.Get(headerKey), nil
+	})
+	// Finish the EntrySpan.
+	defer trace.StopSpan()
+
+	_, _ = w.Write([]byte("success from provider"))
+}
+
+func main() {
+	http.HandleFunc("/provider", providerHandler)
+
+	_ = http.ListenAndServe(":9998", nil)
+}
+
+```
+
+
+
+Then, in the terminal, execute:
+
+```shell
+go build -toolexec="/path/to/go-agent" -a -o consumer ./consumer.go
+./consumer
+```
+
+```shell
+go build -toolexec="/path/to/go-agent" -a -o provider ./provider.go
+./provider
+```
+
+```shell
+curl 127.0.0.1:9999/consumer
+```
+
+At this point, the UI will display the span information you created.
+
+![Trace List](./trace.png)
+
+If you need to trace methods that are executed only locally, you can use trace.CreateLocalSpan(). If you don't need to monitor information or states from the other end, you can change ExitSpan and EntrySpan to LocalSpan.
+
+The usage examples provided are for illustration purposes, and users can decide the tracing granularity and where in the program they need tracing.
+
+Please note that if a program ends too quickly, it may cause tracing data to be unable to be asynchronously sent to the SkyWalking backend.
+
+## Populate The Span
+
+When there's a necessity to record additional information, including creating/updating tags, appending logs, and setting a new operation name of the current traced Span, these APIs should be considered. These actions are used to enhance trace information, providing a more detailed and precise contextual description, which aids in better understanding the events or operations being traced.
+
+Toolkit trace APIs provide a convenient way to access and manipulate trace data, including:
+
+- Setting Tags: `SetTag()`
+- Adding Logs: `AddLog()`
+- Setting Span Names: `SetOperationName()`
+- Getting various IDs: `GetTraceID()`, `GetSegmentID()`, `GetSpanID()`
+
+For example, if you need to record the HTTP status code in a span, you can use the following interfaces while the span is not yet finished:
+
+```go
+trace.CreateExitSpan("GET:/provider", "localhost:9999", func(headerKey, headerValue string) error {
+	r.Header.Add(headerKey, headerValue)
+	return nil
+})
+resp, err := http.Get("http://localhost:9999/provider")
+trace.SetTag("status_code", fmt.Sprintf("%d", resp.StatusCode))
```

Comment:
Stupid question from my side, how do we make sure that concurrency does not interfere with the current trace? Do we keep that in the context?

Reply:
There is [a context propagation mechanism](https://skywalking.apache.org/docs/skywalking-go/next/en/concepts-and-designs/key-principles/#context-propagation-between-methods) to keep the trace. For short, We would obtain real-time trace data from the current goroutine when setting the tag.